ROCm/HIP (#452)
getsolus/help-center-docs: docs/user/software/development/rocm.md

---
title: ROCm/HIP
summary: A quick guide to getting set up for ROCm/HIP development on Solus
---

# ROCm/HIP

ROCm is AMD's open-source software stack for GPU computation.

Note that ROCm is not required in order for, say, your display or browser, to

> use GPU-acclerated rendering. These are more on the driver side of things and 

Reviewer: ```suggestion
use GPU-accelerated rendering. These are more on the driver side of things and 
```
Other than this, spelling checks out

are
handled by the kernel and/or Mesa. ROCm is mainly focused on GPU-accelerated
computing, such as GPU rendering in Blender or GPU-accelerated machine learning
in PyTorch.

## Install ROCm/HIP

> 

Reviewer 2: Add an introductory sentence to this block:

```suggestion
To install ROCm, execute the following command:
```

```bash
sudo eopkg it rocm-hip rocm-opencl
```

If you are also developing with ROCm/HIP, install the
development files and the `hipcc` compiler driver as well:
```bash
sudo eopkg it rocm-hip-devel
```

## Necessary Environment Variables

It is recommended and safe to put these environment variables in your
`~/.bashrc`:
```bash
export ROCM_PATH=/usr
export HIP_PATH=/usr
```

If you're developing with ROCm/HIP, the following environment variables will
save you a lot of hassle:
```bash
export HIP_DEVICE_LIB_PATH=/usr/lib64/amdgcn/bitcode
export DEVICE_LIB_PATH=$HIP_DEVICE_LIB_PATH
export HIP_PLATFORM=amd
export HIP_RUNTIME=amd
export HIP_ROCCLR_HOME=$ROCM_PATH
```

## Supported Hardware and GPU Architectures

<!--
ROCm is designed such that in order to for the compiled binaries to run on a 
certain GPU model, during compiling one must compile with that GPU as the
compilation target. 
!-->

> ROCm is designed such that in order for a library to support N different GPU

Reviewer 2: This is a [wall of text](https://developers.google.com/style/accessibility#ease-of-reading), nobody will read this. 

With that being said, is it even necessary to mention all of this? I mean, for end users what matters is whether the GPU architecture is supported or not: If it's supported, they'll continue their day without reading everything; if it's not supported, they'll reach out using the available support channels.

```suggestion

:::tip

Install and run `rocminfo`  to see your GPU architecture.

:::

The ROCm package included in Solus supports the following GPU architectures out of the box:

- `gfx803`
- `gfx900`
- `gfx906`
- `gfx908`
- `gfx90a`
- `gfx1010`. This architecture can run programs compiled for `gfx101*` GPUs such as `gfx1011` and `gfx1012`, see [Emulating
  as a Supported Architecture](#emulating-as-a-supported-architecture).
- `gfx1030`. This architecture can run programs compiled for `gfx103*` GPUs such as `gfx1031` and `gfx1032`, see [Emulating
  as a Supported Architecture](#emulating-as-a-supported-architecture).
- `gfx1010`
- `gfx1011`
- `gfx1012`

If your GPU is not on the list, please open an issue in our [issue tracker](https://github.com/getsolus/packages/issues/new/choose) with your GPU model and release year.

This list is only the minimum supported architectures. Some packages like [Blender](#blender) support more architectures.
```

architectures, that library must be compiled N times, once for each
architecture, causing the build time of a package to grow linearly. For example,
if we want PyTorch to support running on 5 different GPU architectures, we
essentially need to compile PyTorch 5 times. It should be obvious this quickly
becomes a maintenance burden as the compile time grows linearly with respect to
the number of GPUs models we want to support.

Therefore, we have carefuly picked
the following baseline
architectures such that we support as many reasonably recent
hardware as possible while not causing compilation times to skyrocket. Any GPU
architecture in the list below should work out-of-the-box.

- `gfx803`
- `gfx900`
- `gfx906`
- `gfx908`
- `gfx90a`
- `gfx1010`; for `gfx101*` GPUs such as `gfx1011` and `gfx1012`, see [Emulating
  as a Supported Architecture](#emulating-as-a-supported-architecture) section.
- `gfx1030`; for `gfx103*` GPUs such as `gfx1031` and `gfx1032`, see [Emulating
  as a Supported Architecture](#emulating-as-a-supported-architecture) section.
- `gfx1010`
- `gfx1011`
- `gfx1012`

> :::tip

Reviewer 2: Move this note to the top of the section (See my first comment for more info)

Run `rocminfo` provided by the `rocminfo` package to 
see what architecture your GPU(s) have.

:::

> :::note

Reviewer 2: This shouldn't be a note (See my first comment for more info)

This list is only the minimum supported architectures. Some packages like
[Blender](#blender) are built with support for even more architectures.

:::

> If your GPU model is not on the list, please open an issue in

Reviewer 2: Move this right after the GPU list.

our [Issue Tracker] with your GPU model and the year that this model is
released.

### Emulating as a Supported Architecture

Several GPU archiectures, such as those in the Navi 1 family, have

> almost identical (if not exactly identical) ISA that allows a program compiled for

Reviewer 2: Simplify this line:
```suggestion
similar ISAs. This allows programs compiled for
```

> one architecture to run seamlessly on other. 

Reviewer 2: [Avoid excessive claims](https://developers.google.com/style/excessive-claims)
```suggestion
one architecture to run on another. 
```

For example, any program compiled for the `gfx1030` architecture can also run on

> the `gfx1031` and `gfx1032` architectures. A list of such architectures is

Reviewer 2: In general, it's not a good idea to say things like "the previous section" or "the previous step". Instead, refer to the specific section name and include a link to it.

```suggestion
the `gfx1031` and `gfx1032` architectures. See the _Supported Hardware and GPU Architectures_ list for the supported architectures.
```

listed in the previous section.

> To emulate your GPU as a supported architecture, the environment variable

Reviewer 2: [Avoid passive voice](https://developers.google.com/style/voice):

```suggestion
To emulate your GPU as a supported architecture, add the `HSA_OVERRIDE_GFX_VERSION` environment variable to your system. For example:
```
